File: src/main/java/org/manapart/unyeilding_items/UnyieldingItems.java
package org.manapart.unyeilding_items;

import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.util.Hand;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.entity.living.LivingHurtEvent;
import net.minecraftforge.event.entity.player.*;
import net.minecraftforge.eventbus.api.IEventBus;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent;
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

// The value here should match an entry in the META-INF/mods.toml file
@Mod("unyielding_items")
public class UnyieldingItems {

    private static final Logger LOGGER = LogManager.getLogger();

    public UnyieldingItems() {
        MinecraftForge.EVENT_BUS.register(this);
    }

    @SubscribeEvent
    public void onRightClickItem(PlayerInteractEvent.RightClickItem event) {
        repairItemInHand(event.getPlayer(), event.getHand());
    }

    @SubscribeEvent
    public void onLeftClickBlock(PlayerInteractEvent.LeftClickBlock event) {
        repairItemInHand(event.getPlayer(), event.getHand());
    }

    @SubscribeEvent
    public void onHoeUse(UseHoeEvent event) {
        repairItemInHand(event.getPlayer(), Hand.MAIN_HAND);
    }

    @SubscribeEvent
    public void onPickup(EntityItemPickupEvent event) {
        repairItem(event.getItem().getItem());
    }

    @SubscribeEvent
    public void onDestroy(PlayerDestroyItemEvent event) {
        ItemStack item = event.getOriginal();
        if (shouldRepair(item)) {
            item.getItem().setDamage(item, 1);
            event.getPlayer().addItem(item);
        }
    }

    @SubscribeEvent
    public void onHurt(LivingHurtEvent event) {
        if (event.getEntityLiving() instanceof PlayerEntity) {
            PlayerEntity player = (PlayerEntity) event.getEntityLiving();
            repairAll(player);
        }
    }

    private void repairAll(PlayerEntity player) {
        repairItemInHand(player, Hand.MAIN_HAND);
        repairItemInHand(player, Hand.OFF_HAND);
        for (ItemStack stack : player.getArmorSlots()) {
            repairItem(stack);
        }
    }

    private void repairItemInHand(PlayerEntity player, Hand hand) {
        repairItem(player.getItemInHand(hand));
    }

    private void repairItem(ItemStack stack) {
        if (shouldRepair(stack)) {
//            System.out.println("Healing " + stack.getDisplayName() + " on swing");
            stack.getItem().setDamage(stack, 1);
        }
    }

    private boolean shouldRepair(ItemStack stack) {
        if (stack != null) {
            return stack.getItem().isDamageable(stack) || stack.getItem().isRepairable(stack);
        }
        return false;
    }

}

File: src/main/java/org/manapart/unyeilding_items/UnyieldingItems.kt
package org.manapart.unyeilding_items

import net.minecraft.world.InteractionHand
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraftforge.event.entity.living.LivingHurtEvent
import net.minecraftforge.event.entity.player.EntityItemPickupEvent
import net.minecraftforge.event.entity.player.PlayerDestroyItemEvent
import net.minecraftforge.event.entity.player.PlayerInteractEvent.LeftClickBlock
import net.minecraftforge.event.entity.player.PlayerInteractEvent.RightClickItem
//import net.minecraftforge.event.entity.player.UseHoeEvent
import net.minecraftforge.fml.common.Mod
import thedarkcolour.kotlinforforge.forge.FORGE_BUS

@Mod("unyielding_items")
class UnyieldingItems {

    init {
        FORGE_BUS.addListener { event: RightClickItem -> repairItemInHand(event.player, event.hand) }
        FORGE_BUS.addListener { event: LeftClickBlock -> repairItemInHand(event.player, event.hand) }
        FORGE_BUS.addListener { event: EntityItemPickupEvent -> repairItem(event.item.item) }
        FORGE_BUS.addListener(::onDestroy)
        FORGE_BUS.addListener(::onHurt)
    }

    fun onDestroy(event: PlayerDestroyItemEvent) {
        val item = event.original
        if (shouldRepair(item)) {
            item.item.setDamage(item, 1)
            event.player.addItem(item)
        }
    }

    fun onHurt(event: LivingHurtEvent) {
        if (event.entityLiving is Player) {
            val player = event.entityLiving as Player
            repairAll(player)
        }
    }

    private fun repairAll(player: Player) {
        repairItemInHand(player, InteractionHand.MAIN_HAND)
        repairItemInHand(player, InteractionHand.OFF_HAND)
        for (stack in player.armorSlots) {
            repairItem(stack)
        }
    }

    private fun repairItemInHand(player: Player, hand: InteractionHand) {
        repairItem(player.getItemInHand(hand))
    }

    private fun repairItem(stack: ItemStack) {
        if (shouldRepair(stack)) {
//            System.out.println("Healing " + stack.getDisplayName() + " on swing");
            stack.item.setDamage(stack, 1)
        }
    }

    private fun shouldRepair(stack: ItemStack?): Boolean {
        return if (stack != null) {
            stack.item.isDamageable(stack) || stack.item.isRepairable(stack)
        } else false
    }
}